File: LeetCode-Java/Trapping Rain Water.java
public class Solution {
    public int trap(int[] A) {
        int left = 0,right = A.length - 1;
        int result = 0;
        while(left < right)
        {
            while(left < right && A[left] <= A[left + 1]) left ++;
            while(right > left && A[right] <= A[right - 1]) right --;
            if(left >= right) break;
            if(A[left] < A[right])
            {
                int startHeight = A[left];
                left ++;
                while(left < right && A[left] < startHeight )
                {
                    result += startHeight - A[left];
                    left ++;
                }
            }
            else
            {
                int startHeight = A[right];
                right --;
                while(right > left && A[right] < startHeight)
                {
                    result += startHeight - A[right];
                    right --;
                }
            }
        }
        return result;
    }
}

File: Leetcode-cpp/trapping_rain_water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int result = 0;
        stack<pair<int ,int>> record;
        for (int i = 0; i < height.size(); ++i) {
            int record_pos = i;
            while (!record.empty() && record.top().second < height[i]) {
                int pos = record.top().first;
                int val = record.top().second;
                record.pop();
                if (record.empty()) break;
                result += (i - pos) * (min(height[i], record.top().second) - val);
                record_pos = pos;
            }
            record.push(make_pair(record_pos, height[i]));
        }
        return result;
    }
};